tendermint/src/application.rs

```diff
@@ -110,109 +115,200 @@ pub trait ABCIApplication<G>: Send + Clone + 'static {
 }
 
 #[derive(Debug, Clone)]
-pub struct ABCI<T: ABCIApplication<G>, G> {
+pub struct ABCI<T: ABCIApplication<G, C>, C: Clone + Send + Sync, G> {
     handler: T,
+    token: CancellationToken<C>,
     _phantom: std::marker::PhantomData<G>,
 }
 
-impl<G, T: ABCIApplication<G>> From<T> for ABCI<T, G> {
+impl<G, C: Clone + Send + Sync, T: ABCIApplication<G, C>> From<T> for ABCI<T, C, G> {
     fn from(handler: T) -> Self {
         Self {
             handler,
             _phantom: Default::default(),
+            token: CancellationToken::new(),
         }
     }
 }
 
-impl<G: DeserializeOwned + Send + Clone + 'static, T: ABCIApplication<G>> Application
-    for ABCI<T, G>
+impl<
+        G: DeserializeOwned + Send + Clone + 'static,
+        C: Clone + Send + Sync + 'static,
+        T: ABCIApplication<G, C>,
+    > Application for ABCI<T, C, G>
 {
     fn echo(
         &self,
         request: tendermint_proto::abci::RequestEcho,
     ) -> tendermint_proto::abci::ResponseEcho {
+        self.token.panic_if_cancelled();
+
         T::echo(&self.handler, request.into()).into()
     }
 
     fn info(
         &self,
         request: tendermint_proto::abci::RequestInfo,
     ) -> tendermint_proto::abci::ResponseInfo {
+        self.token.panic_if_cancelled();
+
         T::info(&self.handler, request.into()).into()
     }
 
     fn init_chain(
         &self,
         request: tendermint_proto::abci::RequestInitChain,
     ) -> tendermint_proto::abci::ResponseInitChain {
-        T::init_chain(&self.handler, request.try_into().unwrap_or_invalid()).into()
+        self.token.panic_if_cancelled();
+
+        let guard = self.token.drop_guard();
+
+        let result = T::init_chain(
+            &self.handler,
+            request.try_into().unwrap_or_invalid(),
+            self.token.clone(),
+        );
+
+        guard.disarm();
+
+        result.into()
     }
 
     fn query(
         &self,
         request: tendermint_proto::abci::RequestQuery,
     ) -> tendermint_proto::abci::ResponseQuery {
-        T::query(&self.handler, request.into()).into()
+        self.token.panic_if_cancelled();
+        let guard = self.token.drop_guard();
+
+        let result = T::query(&self.handler, request.into(), self.token.clone());
+
+        guard.disarm();
+
+        result.into()
     }
 
     fn check_tx(
         &self,
         request: tendermint_proto::abci::RequestCheckTx,
     ) -> tendermint_proto::abci::ResponseCheckTx {
-        T::check_tx(&self.handler, request.into()).into()
+        match self.token.if_cancelled_ctx() {
+            Some(ctx) => match ctx {
+                CancellationContext::GasOverflow => {
+                    ResponseCheckTx::error_with_gas_overflow().into()
+                }
+                _ => {
+                    self.token.panic_if_cancelled();
```

Comment:
What are the other cases here? It's a little unclear here.

Reply:
Whatever you like. This code only to show possible usage for gas and maybe something else

Reply:
> Whatever you like. This code only to show possible usage for gas and maybe something else

Okay. But improved version may have concrete variants except gas metering and it's better to add them.

Reply:
I'm unsure that we need gas in token...